### PR TITLE
StateProcessing Speedup

### PR DESCRIPTION
**Pull request by jrhea**
## PR Description
Added basic framework for caching variables that need to be calculated many times in a slot
## Fixed Issue(s)

- shuffling only needs to be called 1x per slot instead of 2 or 3 times (depending on it if it is an epoch boundary)
- made the parameterless constructor in BeaconState responsible for initializing it's own variables to valid values (removed that logic from get_initial_beacon_state)
- reimplemented DeepCopy() without serialization
- fix to retrieving latest_crosslinks to ensure that we don't index a value outside of the range.

### CODE REVIEW

File: ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/ValidatorsUtil.java [hunk 1]
```diff
@@ -62,6 +62,7 @@
       List<Validator> validators, UnsignedLong epoch) {
     List<Integer> active_validator_indices = new ArrayList<>();
     IntStream.range(0, validators.size())
+        .parallel()
```
Comment (cemozerr) [inline ethereum/datastructures/src/main/java/tech/pegasys/artemis/datastructures/util/ValidatorsUtil.java:65]: 🔥

File: ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/EpochProcessorUtil.java [hunk 2]
```diff
@@ -444,17 +444,22 @@ public static void attestionInclusion(BeaconState state, UnsignedLong previous_t
...
-      }
+      previous_indices
+          .parallelStream()
+          .forEach(
+              index -> {
+                UnsignedLong inclusion_slot = AttestationUtil.inclusion_slot(state, index);
+                int proposer_index =
+                    BeaconStateUtil.get_beacon_proposer_index(state, inclusion_slot);
+                List<UnsignedLong> balances = state.getValidator_balances();
+                UnsignedLong balance = balances.get(proposer_index);
```
Comment (cemozerr) [inline ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/EpochProcessorUtil.java]: Does Java Stream API handle race conditions that might arise when parallel processing the balance for the same proposer_index in two different streams? or should we be worried about that?

Comment (jrhea) [reply ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/EpochProcessorUtil.java]: I need to look through these again. In general, there is no magic bullet. You should always make sure that the state of one element does not affect another element.

Comment (cemozerr) [reply ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/util/EpochProcessorUtil.java]: 👍

**Review by cemozerr — APPROVED** [review]
LGTM.

**Pull request merged**